# Dataset Name - New York City Taxi Trips 

### Extracting Json Data from MongoDB and Storing it in a Data Frame - 'df_nyc_taxi'

In [3]:
#importing library 
import pymongo
import pandas as pd
from pandas import DataFrame

#Connecting to local MongoDB database 
client = pymongo.MongoClient("mongodb://localhost:27017/")

# Database Name
db_name = client["DAP_Project"]

# Collection Name
collection = db_name["New_York_Taxi"]

#Extarcting and Storing the data in the Data Frame
df_nyc_taxi = DataFrame(list(collection.find({})))
df_nyc_taxi

_id vendor_id                   pickup_datetime  \
0       6256c0ac584088e1cafaf38f       CMT  2012-06-09T16:21:13.580003+00:00   
1       6256c0ac584088e1cafaf390       CMT  2012-08-14T21:44:14.940659+00:00   
2       6256c0ac584088e1cafaf391       VTS  2012-07-06T17:12:07.862531+00:00   
3       6256c0ac584088e1cafaf392       VTS  2012-07-22T04:12:28.704958+00:00   
4       6256c0ac584088e1cafaf393       VTS   2012-07-11T11:40:13.77461+00:00   
...                          ...       ...                               ...   
999995  6256c18a584088e1ca0a35cb       CMT   2012-07-02T02:16:07.72331+00:00   
999996  6256c18a584088e1ca0a35cc       CMT  2012-04-11T16:24:49.255035+00:00   
999997  6256c18a584088e1ca0a35cd       CMT   2012-09-15T15:00:58.80813+00:00   
999998  6256c18a584088e1ca0a35ce       CMT   2012-06-25T09:17:36.81712+00:00   
999999  6256c18a584088e1ca0a35cf       CMT  2012-08-24T07:25:13.503063+00:00   

                        dropoff_datetime  passenger_count  trip_distance  \
0       2012-06-09T16:31:19.862787+00:00                2           0.80   
1        2012-08-14T21:54:50.01863+00:00                1           5.40   
2       2012-07-06T17:22:57.404442+00:00                5           1.67   
3       2012-07-22T04:20:45.142945+00:00                1           4.13   
4       2012-07-11T11:49:03.034108+00:00                1           2.03   
...                                  ...              ...            ...   
999995   2012-07-02T02:24:36.05472+00:00                3           6.30   
999996  2012-04-11T16:30:46.179091+00:00                1           1.80   
999997  2012-09-15T15:08:10.313917+00:00                4           2.90   
999998  2012-06-25T09:31:07.322351+00:00                2           1.70   
999999  2012-08-24T07:38:33.870016+00:00                1           2.00   

        pickup_longitude  pickup_latitude rate_code  store_and_fwd_flag  \
0             -74.004114        40.742950      None                 NaN   
1             -73.996506        40.747784      None                 NaN   
2             -73.951340        40.752070      None                 NaN   
3             -74.001117        40.729128      None                 NaN   
4             -73.976600        40.756873      None                 NaN   
...                  ...              ...       ...                 ...   
999995        -73.947063        40.784441      None                 NaN   
999996          0.000000         0.000000      None                 NaN   
999997        -74.006156        40.743763      None                 NaN   
999998        -73.979615        40.783816      None                 NaN   
999999        -73.988829        40.731162      None                 NaN   

        dropoff_longitude  dropoff_latitude payment_type  fare_amount  \
0              -73.994712         40.747950         Cash          5.4   
1              -73.940449         40.792385         Cash         15.4   
2              -73.944535         40.768108         CASH          6.1   
3              -73.958122         40.721812         CASH         12.9   
4              -73.979757         40.776075         CASH          7.7   
...                   ...               ...          ...          ...   
999995         -73.987610         40.738477         Cash         19.4   
999996           0.000000          0.000000         Cash          7.4   
999997         -73.974005         40.764069         Cash         10.6   
999998         -73.988470         40.764176       Credit          7.4   
999999         -74.009790         40.710155         Cash          7.8   

        surcharge  tip_amount  tolls_amount  total_amount  
0             0.0         0.0           0.0           5.4  
1             0.0         0.0           0.0          15.4  
2             1.0         0.0           0.0           7.1  
3             0.5         0.0           0.0          13.4  
4             0.5         0.0           0.0           8.2  
...           ...

### Initial Overview of the Data

##### Checking the number of rows and columns in the dataset

In [70]:
row_and_columns = list(df_nyc_taxi.shape)


print("\nNumber of Rows in the Dataset : ", row_and_columns[0])
print("Number of Columns in the Dataset : ", row_and_columns[1])


Number of Rows in the Dataset :  1000000
Number of Columns in the Dataset :  18


##### Exploring the Columns/Features of the Dataset

In [71]:
column_names = list(df_nyc_taxi.columns)

print("\nThe Dataset contains following Columns :")
for column in column_names:
    print(column)
 
print("\nStatistics of the Data is as follows :")
df_nyc_taxi.describe()


The Dataset contains following Columns :
_id
vendor_id
pickup_datetime
dropoff_datetime
passenger_count
trip_distance
pickup_longitude
pickup_latitude
rate_code
store_and_fwd_flag
dropoff_longitude
dropoff_latitude
payment_type
fare_amount
surcharge
tip_amount
tolls_amount
total_amount

Statistics of the Data is as follows :


passenger_count   trip_distance  pickup_longitude  pickup_latitude  \
count   1000000.000000  1000000.000000    1000000.000000   1000000.000000   
mean          1.733573        2.678209        -72.830595        40.124727   
std           1.254536        3.130602          9.122398         5.026364   
min           0.000000        0.000000        -84.878351        -0.009262   
25%           1.000000        1.000000        -73.991124        40.736387   
50%           1.000000        1.700000        -73.981110        40.754531   
75%           2.000000        3.000000        -73.966815        40.768442   
max           6.000000       49.920000          0.004023        47.922847   

       store_and_fwd_flag  dropoff_longitude  dropoff_latitude  \
count                62.0     1000000.000000    1000000.000000   
mean                  0.0         -72.848044         40.134998   
std                   0.0           9.050547          4.986515   
min                   0.0         -79.191394         -0.015202   
25%                   0.0         -73.990665         40.735366   
50%                   0.0         -73.979534         40.754488   
75%                   0.0         -73.963385         40.769523   
max                   0.0           0.008668         47.935812   

          fare_amount       surcharge      tip_amount    tolls_amount  \
count  1000000.000000  1000000.000000  1000000.000000  1000000.000000   
mean         9.629421        0.161939        0.408540        0.123594   
std          7.567481        0.308355        1.258633        0.761664   
min          2.500000        0.000000        0.000000        0.000000   
25%          5.400000        0.000000        0.000000        0.000000   
50%          7.300000        0.000000        0.000000        0.000000   
75%         10.900000        0.000000        0.000000        0.000000   
max        200.000000        1.000000       98.230000       20.000000   

         total_amount  
count  1000000.000000  
mean        10.326866  
std          8.503211  
min          2.500000  
25%          5.700000  
50%          7.800000  
75%         11.400000  
max        230.000000

##### Checking the datatypes of the Columns

In [72]:
print("\nDatatypes of the Columns")
print('--------------------------------')
print("\nColumns with 'int' datatype : ")
for i in list(df_nyc_taxi.select_dtypes(include=['int64']).columns):
    print(i)

print('\n--------------------------------')
print("Columns with 'float' datatype : ")
for f in list(df_nyc_taxi.select_dtypes(include=['float64']).columns):
    print(f)

print('\n--------------------------------')
print("Columns with 'String' datatype : ")
for s in list(df_nyc_taxi.select_dtypes(include=['object']).columns):
    print(s)


Datatypes of the Columns
--------------------------------

Columns with 'int' datatype : 
passenger_count

--------------------------------
Columns with 'float' datatype : 
trip_distance
pickup_longitude
pickup_latitude
store_and_fwd_flag
dropoff_longitude
dropoff_latitude
fare_amount
surcharge
tip_amount
tolls_amount
total_amount

--------------------------------
Columns with 'String' datatype : 
_id
vendor_id
pickup_datetime
dropoff_datetime
rate_code
payment_type


As seen the above results, the columns - **pickup_datetime** and  **dropoff_datetime** falls under the data-type of *string*, which technically should fall under the *datetime*.

Also all the instances in these two columns have a string 'T' in them, which should be removed first removed before converting the data type to datetime

##### Removing the string ' T '   from the instances of column - 'pickup_datetime' and 'dropoff_datetime' and changing their data-type to ' datetime ' 

In [73]:
# Removing 'T' from the columns - 'pickup_datetime' and 'dropoff_datetime'
df_nyc_taxi['pickup_datetime'] = df_nyc_taxi['pickup_datetime'].replace("T", " ", regex=True)
df_nyc_taxi['dropoff_datetime'] = df_nyc_taxi['dropoff_datetime'].replace("T", " ", regex=True)

df_nyc_taxi.head()

_id vendor_id                   pickup_datetime  \
0  6256c0ac584088e1cafaf38f       CMT  2012-06-09 16:21:13.580003+00:00   
1  6256c0ac584088e1cafaf390       CMT  2012-08-14 21:44:14.940659+00:00   
2  6256c0ac584088e1cafaf391       VTS  2012-07-06 17:12:07.862531+00:00   
3  6256c0ac584088e1cafaf392       VTS  2012-07-22 04:12:28.704958+00:00   
4  6256c0ac584088e1cafaf393       VTS   2012-07-11 11:40:13.77461+00:00   

                   dropoff_datetime  passenger_count  trip_distance  \
0  2012-06-09 16:31:19.862787+00:00                2           0.80   
1   2012-08-14 21:54:50.01863+00:00                1           5.40   
2  2012-07-06 17:22:57.404442+00:00                5           1.67   
3  2012-07-22 04:20:45.142945+00:00                1           4.13   
4  2012-07-11 11:49:03.034108+00:00                1           2.03   

   pickup_longitude  pickup_latitude rate_code  store_and_fwd_flag  \
0        -74.004114        40.742950      None                 NaN   
1        -73.996506        40.747784      None                 NaN   
2        -73.951340        40.752070      None                 NaN   
3        -74.001117        40.729128      None                 NaN   
4        -73.976600        40.756873      None                 NaN   

   dropoff_longitude  dropoff_latitude payment_type  fare_amount  surcharge  \
0         -73.994712         40.747950         Cash          5.4        0.0   
1         -73.940449         40.792385         Cash         15.4        0.0   
2         -73.944535         40.768108         CASH          6.1        1.0   
3         -73.958122         40.721812         CASH         12.9        0.5   
4         -73.979757         40.776075         CASH          7.7        0.5   

   tip_amount  tolls_amount  total_amount  
0         0.0           0.0           5.4  
1         0.0           0.0          15.4  
2         0.0           0.0           7.1  
3         0.0           0.0          13.4  
4         0.0           0.0           8.2

In [74]:
#Changing the data-type of 'pickup_datetime' and 'dropoff_datetime' to datetime

df_nyc_taxi['pickup_datetime'] = pd.to_datetime(df_nyc_taxi['pickup_datetime'], format='%Y-%m-%d %H:%M:%S.%f')
df_nyc_taxi['dropoff_datetime'] = pd.to_datetime(df_nyc_taxi['dropoff_datetime'], format='%Y-%m-%d %H:%M:%S.%f')

In [75]:
#Checking if the data-type as changed
print('pickup_datetime data-type : ', df_nyc_taxi['pickup_datetime'].dtype)
print('dropoff_datetime data-type : ', df_nyc_taxi['dropoff_datetime'].dtype)

pickup_datetime data-type :  datetime64[ns, UTC]
dropoff_datetime data-type :  datetime64[ns, UTC]


##### Investigating for null values

In [76]:
print('Columns and their number of null-values : ')
df_nyc_taxi.shape
for na in df_nyc_taxi.columns:
    print(na,' - ',df_nyc_taxi[na].isna().sum())

Columns and their number of null-values : 
_id  -  0
vendor_id  -  0
pickup_datetime  -  0
dropoff_datetime  -  0
passenger_count  -  0
trip_distance  -  0
pickup_longitude  -  0
pickup_latitude  -  0
rate_code  -  1000000
store_and_fwd_flag  -  999938
dropoff_longitude  -  0
dropoff_latitude  -  0
payment_type  -  0
fare_amount  -  0
surcharge  -  0
tip_amount  -  0
tolls_amount  -  0
total_amount  -  0


All the values in the column **' rate_code '** are **Null**. Similarly for the column **' store_and_fwd_flag '** , 999938 instances out of 1000000 have a **Null value**. So it is better to drop these two columns

In addition to that, the  **'_id '** column in the Dataset holds unique ID's of each row. Since, the ' _id ' column does not hold much relevance in the analysis, it is better to drop it.

In [77]:
#Dropping the '_id' column
df_nyc_taxi = df_nyc_taxi.drop(['_id'], axis = 1)
print(" \n '_id' Column dropped from the dataset")
df_nyc_taxi = df_nyc_taxi.drop(['rate_code'], axis = 1)
print(" \n 'rate_code' Column dropped from the dataset")
df_nyc_taxi = df_nyc_taxi.drop(['store_and_fwd_flag'], axis = 1)
print(" \n 'store_and_fwd_flag' Column dropped from the dataset")

print('\n Number of Columns reaming : ', list(df_nyc_taxi.shape)[1])
df_nyc_taxi

 
 '_id' Column dropped from the dataset
 
 'rate_code' Column dropped from the dataset
 
 'store_and_fwd_flag' Column dropped from the dataset

 Number of Columns reaming :  15


vendor_id                  pickup_datetime  \
0            CMT 2012-06-09 16:21:13.580003+00:00   
1            CMT 2012-08-14 21:44:14.940659+00:00   
2            VTS 2012-07-06 17:12:07.862531+00:00   
3            VTS 2012-07-22 04:12:28.704958+00:00   
4            VTS 2012-07-11 11:40:13.774610+00:00   
...          ...                              ...   
999995       CMT 2012-07-02 02:16:07.723310+00:00   
999996       CMT 2012-04-11 16:24:49.255035+00:00   
999997       CMT 2012-09-15 15:00:58.808130+00:00   
999998       CMT 2012-06-25 09:17:36.817120+00:00   
999999       CMT 2012-08-24 07:25:13.503063+00:00   

                       dropoff_datetime  passenger_count  trip_distance  \
0      2012-06-09 16:31:19.862787+00:00                2           0.80   
1      2012-08-14 21:54:50.018630+00:00                1           5.40   
2      2012-07-06 17:22:57.404442+00:00                5           1.67   
3      2012-07-22 04:20:45.142945+00:00                1           4.13   
4      2012-07-11 11:49:03.034108+00:00                1           2.03   
...                                 ...              ...            ...   
999995 2012-07-02 02:24:36.054720+00:00                3           6.30   
999996 2012-04-11 16:30:46.179091+00:00                1           1.80   
999997 2012-09-15 15:08:10.313917+00:00                4           2.90   
999998 2012-06-25 09:31:07.322351+00:00                2           1.70   
999999 2012-08-24 07:38:33.870016+00:00                1           2.00   

        pickup_longitude  pickup_latitude  dropoff_longitude  \
0             -74.004114        40.742950         -73.994712   
1             -73.996506        40.747784         -73.940449   
2             -73.951340        40.752070         -73.944535   
3             -74.001117        40.729128         -73.958122   
4             -73.976600        40.756873         -73.979757   
...                  ...              ...                ...   
999995        -73.947063        40.784441         -73.987610   
999996          0.000000         0.000000           0.000000   
999997        -74.006156        40.743763         -73.974005   
999998        -73.979615        40.783816         -73.988470   
999999        -73.988829        40.731162         -74.009790   

        dropoff_latitude payment_type  fare_amount  surcharge  tip_amount  \
0              40.747950         Cash          5.4        0.0         0.0   
1              40.792385         Cash         15.4        0.0         0.0   
2              40.768108         CASH          6.1        1.0         0.0   
3              40.721812         CASH         12.9        0.5         0.0   
4              40.776075         CASH          7.7        0.5         0.0   
...                  ...          ...          ...        ...         ...   
999995         40.738477         Cash         19.4        0.0         0.0   
999996          0.000000         Cash          7.4        0.0         0.0   
999997         40.764069         Cash         10.6        0.0         0.0   
999998         40.764176       Credit          7.4        0.0         2.0   
999999         40.710155         Cash          7.8        0.0         0.0   

        tolls_amount  total_amount  
0                0.0           5.4  
1                0.0          15.4  
2                0.0           7.1  
3                0.0          13.4  
4                0.0           8.2  
...              ...           ...  
999995           0.0          19.4  
999996           0.0           7.4  
999997           0.0          10.6  
999998           0.0           9.4  
999999           0.0           7.8  

[1000000 rows x 15 columns]

Now, the dataset is free from NULL value entries and the data-type of each of the columns is lined as per the feature. 

This *cleaned*, *processed* and *structured* version of the dataset will now be stored in the ***PostreSQL database***, which can be retrived later for further analysis. 


### Storing the Processed and Structured Data in PostgreSQL 

In [58]:
#importing the libraries
from sqlalchemy import create_engine
import psycopg2
import pandas as pd

#Establishing the connection with the local PostreSQL database
connection = create_engine('postgresql://postgres:admin@localhost:5432/dap')

#Sending and Storing the data in the PostgreSQL database
#The name of the table where the dataset will be stored is - NewYorkTaxiTrips
df_nyc_taxi.to_sql('NewYorkTaxiTrips',connection)

### Extracting Data from PostgreSQL 

In [30]:
#Extracting data from PostgreSQL database and strong it in - df_taxi
df_taxi = pd.read_sql('NewYorkTaxiTrips', connection)

In [35]:
df_taxi

vendor_id                  pickup_datetime  \
0            CMT 2012-06-09 16:21:13.580003+00:00   
1            CMT 2012-08-14 21:44:14.940659+00:00   
2            VTS 2012-07-06 17:12:07.862531+00:00   
3            VTS 2012-07-22 04:12:28.704958+00:00   
4            VTS 2012-07-11 11:40:13.774610+00:00   
...          ...                              ...   
999995       CMT 2012-07-02 02:16:07.723310+00:00   
999996       CMT 2012-04-11 16:24:49.255035+00:00   
999997       CMT 2012-09-15 15:00:58.808130+00:00   
999998       CMT 2012-06-25 09:17:36.817120+00:00   
999999       CMT 2012-08-24 07:25:13.503063+00:00   

                       dropoff_datetime  passenger_count  trip_distance  \
0      2012-06-09 16:31:19.862787+00:00                2           0.80   
1      2012-08-14 21:54:50.018630+00:00                1           5.40   
2      2012-07-06 17:22:57.404442+00:00                5           1.67   
3      2012-07-22 04:20:45.142945+00:00                1           4.13   
4      2012-07-11 11:49:03.034108+00:00                1           2.03   
...                                 ...              ...            ...   
999995 2012-07-02 02:24:36.054720+00:00                3           6.30   
999996 2012-04-11 16:30:46.179091+00:00                1           1.80   
999997 2012-09-15 15:08:10.313917+00:00                4           2.90   
999998 2012-06-25 09:31:07.322351+00:00                2           1.70   
999999 2012-08-24 07:38:33.870016+00:00                1           2.00   

        pickup_longitude  pickup_latitude  dropoff_longitude  \
0             -74.004114        40.742950         -73.994712   
1             -73.996506        40.747784         -73.940449   
2             -73.951340        40.752070         -73.944535   
3             -74.001117        40.729128         -73.958122   
4             -73.976600        40.756873         -73.979757   
...                  ...              ...                ...   
999995        -73.947063        40.784441         -73.987610   
999996          0.000000         0.000000           0.000000   
999997        -74.006156        40.743763         -73.974005   
999998        -73.979615        40.783816         -73.988470   
999999        -73.988829        40.731162         -74.009790   

        dropoff_latitude payment_type  fare_amount  surcharge  tip_amount  \
0              40.747950         Cash          5.4        0.0         0.0   
1              40.792385         Cash         15.4        0.0         0.0   
2              40.768108         CASH          6.1        1.0         0.0   
3              40.721812         CASH         12.9        0.5         0.0   
4              40.776075         CASH          7.7        0.5         0.0   
...                  ...          ...          ...        ...         ...   
999995         40.738477         Cash         19.4        0.0         0.0   
999996          0.000000         Cash          7.4        0.0         0.0   
999997         40.764069         Cash         10.6        0.0         0.0   
999998         40.764176       Credit          7.4        0.0         2.0   
999999         40.710155         Cash          7.8        0.0         0.0   

        tolls_amount  total_amount  
0                0.0           5.4  
1                0.0          15.4  
2                0.0           7.1  
3                0.0          13.4  
4                0.0           8.2  
...              ...           ...  
999995           0.0          19.4  
999996           0.0           7.4  
999997           0.0          10.6  
999998           0.0           9.4  
999999           0.0           7.8  

[1000000 rows x 15 columns]

Removing the ' index ' column because, it was not the part of Original Dataset that we initially pushed to the PostgreSQL database, but it came along as the Dataset as result of indexing feature of PostgreSQL database

##### Dropping the ' index ' column

In [32]:
#Droping the index column
df_taxi = df_taxi.drop(['index'], axis = 1)
print(" \n 'index' Column dropped from the dataset")
df_taxi.head()

 
 'index' Column dropped from the dataset


vendor_id                  pickup_datetime                 dropoff_datetime  \
0       CMT 2012-06-09 16:21:13.580003+00:00 2012-06-09 16:31:19.862787+00:00   
1       CMT 2012-08-14 21:44:14.940659+00:00 2012-08-14 21:54:50.018630+00:00   
2       VTS 2012-07-06 17:12:07.862531+00:00 2012-07-06 17:22:57.404442+00:00   
3       VTS 2012-07-22 04:12:28.704958+00:00 2012-07-22 04:20:45.142945+00:00   
4       VTS 2012-07-11 11:40:13.774610+00:00 2012-07-11 11:49:03.034108+00:00   

   passenger_count  trip_distance  pickup_longitude  pickup_latitude  \
0                2           0.80        -74.004114        40.742950   
1                1           5.40        -73.996506        40.747784   
2                5           1.67        -73.951340        40.752070   
3                1           4.13        -74.001117        40.729128   
4                1           2.03        -73.976600        40.756873   

   dropoff_longitude  dropoff_latitude payment_type  fare_amount  surcharge  \
0         -73.994712         40.747950         Cash          5.4        0.0   
1         -73.940449         40.792385         Cash         15.4        0.0   
2         -73.944535         40.768108         CASH          6.1        1.0   
3         -73.958122         40.721812         CASH         12.9        0.5   
4         -73.979757         40.776075         CASH          7.7        0.5   

   tip_amount  tolls_amount  total_amount  
0         0.0           0.0           5.4  
1         0.0           0.0          15.4  
2         0.0           0.0           7.1  
3         0.0           0.0          13.4  
4         0.0           0.0           8.2

## Plots

### Plotting a Bar Chart for Average Number of Trips in a week

In [72]:
week_day = df_taxi['dropoff_datetime'].dt.day_name()
week_day = pd.DataFrame(week_day.value_counts() / 70)
#week_day = week_day.reset_index()
week_day

dropoff_datetime
Tuesday         2059.271429
Wednesday       2052.728571
Saturday        2046.042857
Monday          2045.714286
Friday          2045.600000
Thursday        2044.928571
Sunday          1991.428571

In [73]:
week_day.rename(columns = {'index':'Day_of_the_week', 'dropoff_datetime':'Avg_Number_of_Trips'}, inplace = True)

In [74]:
week_day

Avg_Number_of_Trips
Tuesday            2059.271429
Wednesday          2052.728571
Saturday           2046.042857
Monday             2045.714286
Friday             2045.600000
Thursday           2044.928571
Sunday             1991.428571

In [75]:
sorter = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday']
sorterIndex = dict(zip(sorter,range(len(sorter))))
sorterIndex

{'Sunday': 0,
 'Monday': 1,
 'Tuesday': 2,
 'Wednesday': 3,
 'Thursday': 4,
 'Friday': 5,
 'Saturday': 6}

In [76]:
week_day['Day_id'] = week_day.index
week_day['Day_id'] = week_day['Day_id'].map(sorterIndex)
week_day.head()

Avg_Number_of_Trips  Day_id
Tuesday            2059.271429       2
Wednesday          2052.728571       3
Saturday           2046.042857       6
Monday             2045.714286       1
Friday             2045.600000       5

In [77]:
week_day.sort_values('Day_id', inplace=True)
week_day

Avg_Number_of_Trips  Day_id
Sunday             1991.428571       0
Monday             2045.714286       1
Tuesday            2059.271429       2
Wednesday          2052.728571       3
Thursday           2044.928571       4
Friday             2045.600000       5
Saturday           2046.042857       6

In [78]:
week_day.drop("Day_id", axis=1, inplace=True)
week_day

Avg_Number_of_Trips
Sunday             1991.428571
Monday             2045.714286
Tuesday            2059.271429
Wednesday          2052.728571
Thursday           2044.928571
Friday             2045.600000
Saturday           2046.042857

In [79]:
week_day

Avg_Number_of_Trips
Sunday             1991.428571
Monday             2045.714286
Tuesday            2059.271429
Wednesday          2052.728571
Thursday           2044.928571
Friday             2045.600000
Saturday           2046.042857

In [80]:
week_day.reset_index()

index  Avg_Number_of_Trips
0     Sunday          1991.428571
1     Monday          2045.714286
2    Tuesday          2059.271429
3  Wednesday          2052.728571
4   Thursday          2044.928571
5     Friday          2045.600000
6   Saturday          2046.042857

In [81]:
week_day.rename(columns = {'index':'Day_of_the_week', 'Avg_Number_of_Trips': 'Avg_Number_of_Trips'}, inplace = True)
week_day

Avg_Number_of_Trips
Sunday             1991.428571
Monday             2045.714286
Tuesday            2059.271429
Wednesday          2052.728571
Thursday           2044.928571
Friday             2045.600000
Saturday           2046.042857

In [82]:
week_day = week_day.reset_index()
week_day.rename(columns = {'index':'Day_of_the_week'}, inplace = True)
week_day

Day_of_the_week  Avg_Number_of_Trips
0          Sunday          1991.428571
1          Monday          2045.714286
2         Tuesday          2059.271429
3       Wednesday          2052.728571
4        Thursday          2044.928571
5          Friday          2045.600000
6        Saturday          2046.042857

In [84]:
# bar chart
bar_chart = px.bar(week_day, x='Day_of_the_week', y='Avg_Number_of_Trips', color='Avg_Number_of_Trips', 
                           text_auto='0.2s', title="Avg weekly taxi", height=600)
bar_chart.show()



In [61]:
#Sending and Storing the data in the PostgreSQL database
#The name of the table where the dataset will be stored is - AvgTripsPerWeek

week_day.to_sql('AvgTripsPerWeek',connection)
print("Data successfully stored")

Data successfully stored


### Plotting a Scatter Plot for Trip Distance vs Fare Charged for that distance 

In [17]:
fig = px.scatter(df_taxi, x="trip_distance", y="fare_amount", color="fare_amount",
                  size='trip_distance', hover_data=['fare_amount'])
fig.show()

In [74]:
fig = px.scatter(df_copy, x="trip_distance", y="fare_amount", color="fare_amount",
                  size='trip_distance', hover_data=['fare_amount'])
fig.show()

### Plotting a Line graph for Average Number of Taxi Trips during a particular of the day

In [62]:
hour_taxi_df = pd.DataFrame(df_taxi['dropoff_datetime'].dt.hour.value_counts())

In [63]:
hour_taxi_df

dropoff_datetime
1              41997
15             41991
20             41926
23             41918
4              41907
6              41862
12             41772
21             41765
10             41746
8              41717
7              41663
3              41648
14             41608
5              41596
2              41561
17             41557
18             41552
0              41544
13             41522
11             41508
22             41458
16             41424
19             41397
9              41361

In [64]:
hour_taxi_df['Average_Trips'] = pd.DataFrame(hour_taxi_df['dropoff_datetime'] / 24)

In [65]:
hour_taxi_df

dropoff_datetime  Average_Trips
1              41997    1749.875000
15             41991    1749.625000
20             41926    1746.916667
23             41918    1746.583333
4              41907    1746.125000
6              41862    1744.250000
12             41772    1740.500000
21             41765    1740.208333
10             41746    1739.416667
8              41717    1738.208333
7              41663    1735.958333
3              41648    1735.333333
14             41608    1733.666667
5              41596    1733.166667
2              41561    1731.708333
17             41557    1731.541667
18             41552    1731.333333
0              41544    1731.000000
13             41522    1730.083333
11             41508    1729.500000
22             41458    1727.416667
16             41424    1726.000000
19             41397    1724.875000
9              41361    1723.375000

In [66]:
hour_taxi_df = hour_taxi_df.reset_index()
hour_taxi_df.rename(columns = {'index':'hour(24 hours)', 'dropoff_datetime':'Trips'}, inplace = True)
hour_taxi_df.sort_values(by=['hour(24 hours)'], inplace=True)
hour_taxi_df

hour(24 hours)  Trips  Average_Trips
17               0  41544    1731.000000
0                1  41997    1749.875000
14               2  41561    1731.708333
11               3  41648    1735.333333
4                4  41907    1746.125000
13               5  41596    1733.166667
5                6  41862    1744.250000
10               7  41663    1735.958333
9                8  41717    1738.208333
23               9  41361    1723.375000
8               10  41746    1739.416667
19              11  41508    1729.500000
6               12  41772    1740.500000
18              13  41522    1730.083333
12              14  41608    1733.666667
1               15  41991    1749.625000
21              16  41424    1726.000000
15              17  41557    1731.541667
16              18  41552    1731.333333
22              19  41397    1724.875000
2               20  41926    1746.916667
7               21  41765    1740.208333
20              22  41458    1727.416667
3               23  41918    1746.583333

In [68]:
#import plotly.express as px

fig = px.line(hour_taxi_df, x="hour(24 hours)", y="Average_Trips", title='Average Taxi trips per hour (24 hours)')
fig.show()

In [69]:
#Sending and Storing the data in the PostgreSQL database
#The name of the table where the dataset will be stored is - AvgTripsPerhOUR

hour_taxi_df.to_sql('AvgTripsPerhOUR',connection)
print("Data successfully stored")

Data successfully stored


### Plotting a Scatter Plot for Trip Distance vs Trip Duration

In [70]:
#calculate duration and add a column to the dataframe
from datetime import timedelta

def calculate_duration(row):
    time_difference = (row['dropoff_datetime']-row['pickup_datetime']).total_seconds()
    return (time_difference/60)

df_taxi['duration'] = df_taxi.apply(lambda row: calculate_duration(row), axis=1)


In [71]:
df_taxi.head()

vendor_id                  pickup_datetime                 dropoff_datetime  \
0       CMT 2012-06-09 16:21:13.580003+00:00 2012-06-09 16:31:19.862787+00:00   
1       CMT 2012-08-14 21:44:14.940659+00:00 2012-08-14 21:54:50.018630+00:00   
2       VTS 2012-07-06 17:12:07.862531+00:00 2012-07-06 17:22:57.404442+00:00   
3       VTS 2012-07-22 04:12:28.704958+00:00 2012-07-22 04:20:45.142945+00:00   
4       VTS 2012-07-11 11:40:13.774610+00:00 2012-07-11 11:49:03.034108+00:00   

   passenger_count  trip_distance  pickup_longitude  pickup_latitude  \
0                2           0.80        -74.004114        40.742950   
1                1           5.40        -73.996506        40.747784   
2                5           1.67        -73.951340        40.752070   
3                1           4.13        -74.001117        40.729128   
4                1           2.03        -73.976600        40.756873   

   dropoff_longitude  dropoff_latitude payment_type  fare_amount  surcharge  \
0         -73.994712         40.747950         Cash          5.4        0.0   
1         -73.940449         40.792385         Cash         15.4        0.0   
2         -73.944535         40.768108         CASH          6.1        1.0   
3         -73.958122         40.721812         CASH         12.9        0.5   
4         -73.979757         40.776075         CASH          7.7        0.5   

   tip_amount  tolls_amount  total_amount   duration  
0         0.0           0.0           5.4  10.104713  
1         0.0           0.0          15.4  10.584633  
2         0.0           0.0           7.1  10.825699  
3         0.0           0.0          13.4   8.273966  
4         0.0           0.0           8.2   8.820992

In [73]:
fig = px.scatter(df_taxi, x="trip_distance", y="duration", color="fare_amount",
                  size='trip_distance', hover_data=['fare_amount'])
fig.show()

In [70]:
#Sending and Storing the data in the PostgreSQL database
#The name of the table where the dataset will be stored is - DatawithDuration

hour_taxi_df.to_sql('DatawithDuration',connection)
print("Data successfully stored")

Data successfully stored
